nilearn/plotting/img_plotting.py

```diff
@@ -1982,10 +2018,60 @@ def plot_carpet(img, mask_img=None, detrend=True, output_file=None,
     n_decimations = int(np.ceil(np.log2(np.ceil(n_tsteps / LONG_CUTOFF))))
     data = data[::2 ** n_decimations, :]
 
-    axes.imshow(data.T, interpolation='nearest',
-                aspect='auto', cmap='gray',
-                vmin=vmin or default_vmin,
-                vmax=vmax or default_vmax)
+    if is_atlas:
+        # Define nested GridSpec
+        legend = False
+        wratios = [2, 100, 20]
+        gs = mgs.GridSpecFromSubplotSpec(
+            1,
+            2 + int(legend),
+            subplot_spec=axes,
+            width_ratios=wratios[: 2 + int(legend)],
+            wspace=0.0,
+        )
+
+        ax0 = plt.subplot(gs[0])
+        ax0.set_xticks([])
+        ax0.imshow(
+            atlas_values[:, np.newaxis],
+            interpolation="none",
+            aspect="auto",
+            cmap=cmap
+        )
+        if mask_labels:
+            # Add labels to middle of each associated band
+            mask_labels_inv = {v: k for k, v in mask_labels.items()}
+            ytick_locs = [
+                np.mean(np.where(atlas_values == i)[0])
+                for i in np.unique(atlas_values)
+            ]
+            ax0.set_yticks(ytick_locs)
+            ax0.set_yticklabels([
+                mask_labels_inv[int(i)] for i in np.unique(atlas_values)
```

Comment:
We should have checked earlier that `atlas_values` and `mask_labels` are consistent.

Reply:
I'm having some trouble coming up with a clean way to check this and to coerce the `mask_labels` values to the same datatype as the `atlas_values`.

Reply:
But you can coerce `atlas_values` to int ? It would cover all the use cases I know of ?

Reply:
No, that's just a (bad) placeholder. I plan to have something better working before I switch this PR from draft.

Reply:
@bthirion sorry I completely misread your original comment. I thought you were asking why I did it the way I did instead of recommending changing the target of the datatype change! Your idea was definitely the right approach, while my original approach was the "bad placeholder". 🤦